elixir-fundamentals-review.livemd

```diff
@@ -0,0 +1,339 @@
+
+```elixir
+{a, b, c} = {1, 2, 3}
+{a, b, c}
+```
+
+Use pattern matching to bind the `head` and the `tail` of the following list.
+
+```elixir
+[head | tail] = [1, 2, 3]
+
+tail
+```
+
+Use pattern matching to bind `:a` and `:b` in the following keyword list to variables `a` and `b`.
+
+```elixir
+[a, b] = [a: 1, b: 2]
+[a, b]
+```
+
+## Control Flow
+
+Complete the `Grade` module as documented.
+
+```elixir
+defmodule Grade do
+  @doc """
+  Return the appropriate letter grade for a given integer grade. 
+  Assume you do not need to handle floats.
+
+  86-100: A
+  72-85:  B
+  62-71:  C
+  51-61:  D
+  0-50:   F
+
+  iex> Grade.letter_grade(86)
+  "A"
+  iex> Grade.letter_grade(51)
+  "D"
+  """
+  def letter_grade(integer) do
+    cond do
+      integer >= 86 and integer <= 100 -> "A"
+      integer >= 72 and integer <= 85 -> "B"
+      integer >= 62 and integer <= 71 -> "C"
+      integer >= 51 and integer <= 61 -> "D"
+      true -> "F"
+    end
+  end
+
+  @doc """
+  Return the grade range for a given letter grade.
+
+  A: 86..100
+  B: 72..85
+  C: 62..71
+  D: 51..61
+  F: 0..50
+
+  iex> Grade.grade_range("A")
+  86..100
+  iex> Grade.grade_range("C")
+  62..71
+  """
+  def grade_range(letter) do
+    case letter do
+      "A" -> 86..100
+      "B" -> 72..85
+      "C" -> 62..71
+      "D" -> 51..61
+      _ -> 0..50
+    end
+  end
+
+  @doc """
+  Determine if a number grade is a passing grade (50 or above)
+  iex> Grade.pass?(50)
+  "50 is a passing grade."
+
+  iex> Grade.pass?(100)
+  "100 is a passing grade."
+
+
+  iex> Grade.pass?(20)
+  "20 is a failing grade."
+  """
+  def pass?(integer) do
+    if integer >= 50 do
+      "#{integer} is a passing grade."
+    else
+      "#{integer} is a failing grade."
+    end
+  end
+end
+
+Grade.letter_grade(51)
+Grade.grade_range("C")
+Grade.pass?(100)
+```
+
+## Modules and Structs
+
+In the cell below, define a `Post` struct with `:title` and `:content` keys. Enforce that these keys must be provided a value. You do not need to validate the values.
+
+```mermaid
+erDiagram
+    Post {
+        string title
+        string content
+    }
+```
+
+```elixir
+post = %{title: "book", content: "inside"}
+post
+```
+
+Create an instance of the `Post` struct you defined above.
+
+```elixir
+post
+```
+
+## Enumeration
+
+Complete the `Enumeration` module as documented.
+
+```elixir
+defmodule Enumeration do
+  @doc """
+  Double every integer in a list or range.
+
+  iex> Enumeration.double(1..5)
+  [2, 4, 6, 8, 10]
+  """
+  def double(list) do
+    Enum.map(list, fn each -> each * 2 end)
+  end
+
+  @doc """
+  Return only the even numbers in a list or range.
+
+  iex> Enumeration.evens([2, 2, 5, 4])
+  [2, 2, 4]
+  """
+  def evens(list) do
+    Enum.filter(list, fn each -> rem(each, 2) == 0 end)
+  end
+
+  @doc """
+  Sum only the even numbers in a list or range.
+
+  iex> Enumeration.sum_evens([2, 2, 5, 4])
+  8
+  """
+  def sum_evens(list) do
+    evens(list)
```

Comment:
Excellent job reusing your `evens` function.

Consider avoiding the pipe operator for a single function call. Static analysis tools such as Credo will provide a warning.

You could instead start with the `list` to be more idiomatic.

```elixir
list
|> evens()
|> Enum.sum()
```